Make ValuePlug::dirtyCount Public

danieldresser-ie opened the pull request:
This was pretty hasty since I got caught up in figuring out how to make the Instancer stuff work today, but I went ahead and did a quick naive pass on this.  Two issues:

* When updating the encapsulate tests I realized that we should also make sure changing globals doesn't invalidate capsules.  This now means that the strategy can't be as simple as we suggested in just taking inPlug()->dirtyCount().  I guess we could combine the dirtyCounts of all the child plugs we care about?
* Capsule was explicitly ignoring globalsPlug for the dirtyCount previously, but I've temporarily just naively used the full plug hash in Capsule as well ... and there isn't a test failure, so we should probably both add a test case for this, and fix it.

danieldresser-ie commented:
I've just added the one test for globals not invalidating capsules, changed Capsule and Encapsulate to use the sum of all scenePlug children other than globals, and cleaned up the remaining parentChanged and mutex stuff from Capsule.  Not many changes from yesterday, but the tests are passing now, and it all seems reasonable.

danieldresser-ie commented:
Snuck in one last change - not squashed yet, since I wasn't confident of it.  But after realizing that I would need to use the "sum all dirty counts aside from globals" function in Encapsulate as well, I figured maybe I should make it public.

I'm not sure exactly where it should live ... in theory, it's not specific to capsules.  But we currently have 3 places we need it, all related to Capsules, so this seems like a reasonable place for it.

If this looks good, all I need to do tomorrow is squash it and updates Changes.

johnhaddon commented on python/GafferSceneTest/EncapsulateTest.py:
```diff
@@ -205,6 +205,21 @@ def assertHashesUnique( path ) :
+		c = encapsulate["out"].object( "/group/group" )
+		pathFilter["paths"].setValue( IECore.StringVectorData( [ "/group/group", "/group/group2" ] ) )
+		self.assertEqual( c, encapsulate["out"].object( "/group/group" ) )
+		options["options"].addChild( Gaffer.NameValuePlug( "test", IECore.IntData( 10 ) ) )
+		self.assertEqual( c, encapsulate["out"].object( "/group/group" ) )
```
Sorry, I should have realised this when we chatted before, but capsules actually _do_ [depend on the globals](https://github.com/GafferHQ/gaffer/blob/master/src/GafferScene/Capsule.cpp#L209) at present. We want to fix that somehow so they get rendered with the globals (motion blur etc) from the downstream scene, but that is a battle for another day.

johnhaddon commented on include/Gaffer/ValuePlug.h:
```diff
@@ -205,6 +205,14 @@ class GAFFER_API ValuePlug : public Plug
 
 		//@}
 
+		/// Returns the current dirty count of this plug.
```
I think we need to say what dirty count means, and touch on why we use it for capsules rather than the more usual hash. How about something like this?

```
/// Returns the number of times this plug has been dirtied.
/// This is used internally for cache invalidation but may also
/// be useful for debugging and as part of a "poor man's hash" where
/// computing the full upstream hash might be prohibitively
/// expensive (see `Encapsulate::hashObject()` for example).
```

danieldresser-ie replied:
If we're describing this ... do we need to mention how it's affected by the dirty epoch?

johnhaddon replied:
Ugh, hadn't thought of that, and that'll definitely make debugging a bit more confusing. How do we describe this so it's meaningful without getting into too much detail on the internals? "Returns a value that increments each time the plug is dirtied (but doesn't necessarily start at 0)"?

johnhaddon commented on include/Gaffer/ValuePlug.h:
```diff
@@ -205,6 +205,14 @@ class GAFFER_API ValuePlug : public Plug
+		/// Returns the current dirty count of this plug.
+		/// This is usually only used internally, but is occasionally useful in nodes which have specialized
+		/// ways of storing things that may be invalidated, like Encapsulate
+		uint64_t dirtyCount() const
```
Would be great to add a Python binding for this, as I think it might be quite handy for interactive debugging.

danieldresser-ie replied:
I did this and added tests, however these tests currently do not pass - because I would expect parenting a plug to update the dirtyCount?  This is probably just a case of you telling me why my expectation is wrong, and me updating the test.  But it currently seems quite odd to me that reparenting emits plugSet but doesn't update the dirtyCount, and I probably shouldn't update the test to the current behaviour until I understand why.

johnhaddon replied:
I think you need to parent the plug to a node rather than another plug - we need a node for dirty propagation to be meaningful.

johnhaddon commented on src/GafferScene/Capsule.cpp:
```diff
@@ -84,65 +96,15 @@ Capsule::Capsule(
-
 	if( m_scene )
 	{
-		const_cast<ScenePlug *>( m_scene )->parentChangedSignal().connect(
-			boost::bind( &Capsule::parentChanged, this, ::_1 )
-		);
```
`m_scene` is a raw pointer, and there's a comment in the header saying that we don't need to own a reference because we're using `parentChanged()` to invalidate it instead. Do we need to own a reference now, or should we leave the parent tracking in place for now? Would be nice to get rid of all the signal handling so maybe the reference is the way to go? The choice of raw pointers seems a little odd in the first place, so I'm wondering if my old self was aware of something that I'm not. Hopefully not. If we decide to own a reference, we should also change the constructor argument to `const ConstScenePlugPtr &` to communicate the semantics.

danieldresser-ie replied:
Hmm, I got a bit stumped by this one.  Doing an obvious version of this produces a segfault during exit while destructing the LRUCache.  Is there a limited set of circumstances where it's legal to destruct a ScenePlug?  I guess I'll leave this as a separate commit and see if you can figure it out.  Stack trace looks like this:
```#0  0x00007fffde2b0e28 in tbb::interface6::enumerable_thread_specific<Gaffer::Plug::DirtyPlugs, tbb::cache_aligned_allocator<Gaffer::Plug::DirtyPlugs>, (tbb::ets_key_usage_type)1>::create_local() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#1  0x00007fffde2b1052 in Gaffer::Plug::DirtyPlugs::local() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#2  0x00007fffde2ac9b9 in Gaffer::Plug::pushDirtyPropagationScope() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#3  0x00007fffde2666ac in Gaffer::GraphComponent::~GraphComponent() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#4  0x00007fffde2ade20 in Gaffer::Plug::~Plug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#5  0x00007fffde30283e in Gaffer::ValuePlug::~ValuePlug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#6  0x00007fffc316dc69 in GafferScene::ScenePlug::~ScenePlug() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#7  0x00007fffc306b5de in GafferScene::Capsule::~Capsule() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#8  0x00007fffc306b609 in GafferScene::Capsule::~Capsule() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGafferScene.so
#9  0x00007fffde303572 in IECorePreview::LRUCache<IECore::MurmurHash, boost::intrusive_ptr<IECore::Object const>, IECorePreview::LRUCachePolicy::TaskParallel, (anonymous namespace)::ComputeProcessKey>::~LRUCache() ()
   from /home/danield/apps/gaffer/0.59.0.0dev/cent7.x86_64/cortex/10.1/gaffer/lib/libGaffer.so
#10 0x00007ffff6ae2b69 in __run_exit_handlers () from /lib64/libc.so.6
#11 0x00007ffff6ae2bb7 in exit () from /lib64/libc.so.6
#12 0x00007ffff78c8e1f in Py_Exit (sts=sts@entry=0) at Python/pythonrun.c:1774```

johnhaddon replied:
I think we're just looking at a classic static destruction order fiasco. The cache is static, and so is the `g_dirtyPlugs` used to track plug dirtiness. `g_dirtyPlugs` has been destroyed already, and when the cache is destroyed it deletes a plug which then tries to access `g_dirtyPlugs`. I think our options are to stick with the raw pointer and parent tracking, or to do a tactical "leak" of `g_dirtyPlugs` by declaring it as `static tbb::enumerable_thread_specific<Plug::DirtyPlugs> *g_dirtyPlugs = new ...`;

What we really want is weak pointer support in IECore::RefCounted, or a derived class of it. I've been thinking a bit about adding that as I think there are a few use cases for it, but that'll be a longer term thing.

danieldresser-ie commented:
I've either addressed or commented on the feedback - there are a few things that actually ended up trickier than expected and I'm hoping you have thoughts.  The globalsPlug thing is resolved, but is a bit ugly - we decided to match the old behaviour, which means for the moment, Encapsulate can't use capsuleDirtyCount ( because it would be a problem if we failed to update the hash of Encapsulate when the shutter changes ), but it sounds like we have a plan where all of this cleans up once we address the Capsule shutter thing properly.

danieldresser-ie commented:
Issues addressed.  Parent the plug to a node does make the test make sense, and "leaking" g_dirtyPlugs fixes the crash.  Should be ready for merge now.

johnhaddon commented:
Thanks Daniel! I'm now worrying about the other cause of Capsule destruction though - eviction from the cache during regular operation. I think it's unlikely that the Capsule is the last holder of a reference to the scene in that case, but I'm not sure I want to tempt fate, since it would equate to making a graph edit from inside a compute. Can we keep the use of the dirty count but revert to the original signal-based code for handling parent? I know that has problems too, but I'm thinking maybe it's a case of better-the-devil-you-know. Any thoughts?

danieldresser-ie commented:
As discussed this morning, best plan we've come up with so far is simplying removing invalidation based on dirtyCount altogether, but going back to the old signal-based handling of parent.

danieldresser-ie commented:
As discussed, I've now updated this to just throw out dirtyCount based invalidation, simplifying things, and no longer requiring the leaking of g_dirtyPlugs.

I've renamed it to Capsule::throwIfNoScene, since it doesn't really track expiry any more.
